apps/btc_family/btc_txn.c

```diff
@@ -0,0 +1,286 @@
+ *
+ * @param query Reference to storage for decoding query from host
+ *
+ * @return bool Indicating if the function actions succeeded or failed
+ */
+static bool fetch_transaction_meta(btc_query_t *query);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static txn_context *g_txn_context = NULL;
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+static bool check_which_request(const btc_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->sign_txn.which_request) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_request_data(const btc_sign_txn_request_t *request) {
+  bool status = true;
+
+  if (!btc_derivation_path_guard(request->initiate.derivation_path,
+                                 request->initiate.derivation_path_count)) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+  return status;
+}
+
+static void send_response(const pb_size_t which_response) {
+  btc_result_t result = init_btc_result(BTC_RESULT_SIGN_TXN_TAG);
+  result.sign_txn.which_response = which_response;
+  btc_send_result(&result);
+}
+
+static bool handle_initiate_query(const btc_query_t *query) {
+  char wallet_name[NAME_SIZE] = "";
+  char msg[100] = "";
+
+  // TODO: Handle wallet search failures - eg: Wallet ID not found, Wallet
+  // ID found but is invalid/locked wallet
+  if (!check_which_request(query, BTC_SIGN_TXN_REQUEST_INITIATE_TAG) ||
+      !validate_request_data(&query->sign_txn) ||
+      !get_wallet_name_by_id(query->sign_txn.initiate.wallet_id,
+                             (uint8_t *)wallet_name)) {
+    return false;
+  }
+
+  snprintf(msg, sizeof(msg), UI_TEXT_BTC_SEND_PROMPT, g_app->name, wallet_name);
+  // Take user consent to sign transaction for the wallet
+  if (!core_confirmation(msg, btc_send_error)) {
+    return false;
+  }
+
+  core_status_set_flow_status(BTC_SIGN_TXN_STATUS_CONFIRM);
+  memcpy(&g_txn_context->init_info,
+         &query->sign_txn.initiate,
+         sizeof(btc_sign_txn_initiate_request_t));
+  send_response(BTC_SIGN_TXN_RESPONSE_CONFIRMATION_TAG);
+  return true;
+}
+
+static bool fetch_transaction_meta(btc_query_t *query) {
+  if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) ||
+      !check_which_request(query, BTC_SIGN_TXN_REQUEST_META_TAG)) {
```

Comment:
Any check required to validate the meta data? 

I am guessing we will put a limit on # of inputs and outputs.. so ideally it should go here

Reply:
this is added in #332. can add a todo here